src/proof.rs

```diff
@@ -0,0 +1,1524 @@
+            ZKProofCommitmentField::W_1 => write!(f, "W_1"),
+            ZKProofCommitmentField::W_2 => write!(f, "W_2"),
+            ZKProofCommitmentField::W_3 => write!(f, "W_3"),
+            ZKProofCommitmentField::W_4 => write!(f, "W_4"),
+            ZKProofCommitmentField::Z_PERM => write!(f, "Z_PERM"),
+            ZKProofCommitmentField::LOOKUP_INVERSES => write!(f, "LOOKUP_INVERSES"),
+            ZKProofCommitmentField::LOOKUP_READ_COUNTS => write!(f, "LOOKUP_READ_COUNTS"),
+            ZKProofCommitmentField::LOOKUP_READ_TAGS => write!(f, "LOOKUP_READ_TAGS"),
+            ZKProofCommitmentField::LIBRA_COMMITMENTS(i) => write!(f, "LIBRA_COMMITMENTS_{i}"),
+            ZKProofCommitmentField::GEMINI_FOLD_COMMS(i) => write!(f, "GEMINI_FOLD_COMMS_{i}"),
+            ZKProofCommitmentField::KZG_QUOTIENT => write!(f, "KZG_QUOTIENT"),
+        }
+    }
+}
+
+#[derive(Debug, Hash, Eq, PartialEq)]
+pub enum ProofCommitmentField {
+    SHPLONK_Q,
+    W_1,
+    W_2,
+    W_3,
+    W_4,
+    Z_PERM,
+    LOOKUP_INVERSES,
+    LOOKUP_READ_COUNTS,
+    LOOKUP_READ_TAGS,
+    GEMINI_FOLD_COMMS(usize),
+    KZG_QUOTIENT,
+}
+
+impl fmt::Display for ProofCommitmentField {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ProofCommitmentField::SHPLONK_Q => write!(f, "SHPLONK_Q"),
+            ProofCommitmentField::W_1 => write!(f, "W_1"),
+            ProofCommitmentField::W_2 => write!(f, "W_2"),
+            ProofCommitmentField::W_3 => write!(f, "W_3"),
+            ProofCommitmentField::W_4 => write!(f, "W_4"),
+            ProofCommitmentField::Z_PERM => write!(f, "Z_PERM"),
+            ProofCommitmentField::LOOKUP_INVERSES => write!(f, "LOOKUP_INVERSES"),
+            ProofCommitmentField::LOOKUP_READ_COUNTS => write!(f, "LOOKUP_READ_COUNTS"),
+            ProofCommitmentField::LOOKUP_READ_TAGS => write!(f, "LOOKUP_READ_TAGS"),
+            ProofCommitmentField::GEMINI_FOLD_COMMS(i) => {
+                write!(f, "GEMINI_FOLD_COMMS_{i}")
+            }
+            ProofCommitmentField::KZG_QUOTIENT => write!(f, "KZG_QUOTIENT"),
+        }
+    }
+}
+
+/// Utility function for "reassembling" a `G1ProofPoint` into a `G1`.
+pub(crate) fn convert_proof_point<H: CurveHooks>(
+    g1_proof_point: G1ProofPoint,
+) -> Result<G1<H>, GroupError> {
+    const N: u32 = 136;
+    let x = Fq::from_bigint(g1_proof_point.x_0.bitor(g1_proof_point.x_1.shl(N)))
+        .expect("Should always succeed");
+    let y = Fq::from_bigint(g1_proof_point.y_0.bitor(g1_proof_point.y_1.shl(N)))
+        .expect("Should always succeed");
+
+    if x == Fq::ZERO && y == Fq::ZERO {
+        return Ok(G1::<H>::identity());
+    }
+
+    let point = G1::<H>::new_unchecked(x, y);
+
+    if !point.is_on_curve() {
+        return Err(GroupError::NotOnCurve);
+    }
+
+    // The following cannot happen for G1 with the BN254 curve.
+    // if !point.is_in_correct_subgroup_assuming_on_curve() {...}
```

Comment:
I'd suggest to either delete the commented check, or replace it with a `debug_assert`
```suggestion
    // This is always true for G1 with the BN254 curve.
    debug_assert!(point.is_in_correct_subgroup_assuming_on_curve());
```